File: PHRPReader/clsPepToProteinMapInfo.vb
Option Strict On

Public Class clsPepToProteinMapInfo

    Public Structure udtProteinLocationInfo
        Public ResidueStart As Integer
        Public ResidueEnd As Integer
    End Structure

    ''' <summary>
    ''' Dictionary of protein names and residue start/end positions
    ''' </summary>
    ''' <remarks></remarks>
    Protected mProteinMapInfo As Dictionary(Of String, List(Of udtProteinLocationInfo))

    Public ReadOnly Property ProteinMapInfo As Dictionary(Of String, List(Of udtProteinLocationInfo))
        Get
            Return mProteinMapInfo
        End Get
    End Property

    Public Sub New(ByVal proteinName As String, ByVal residueStart As Integer, ByVal residueEnd As Integer)

        mProteinMapInfo = New Dictionary(Of String, List(Of udtProteinLocationInfo))(StringComparer.CurrentCultureIgnoreCase)

        AddProtein(proteinName, residueStart, residueEnd)
    End Sub

    Public Sub AddProtein(ByVal proteinName As String, ByVal residueStart As Integer, ByVal residueEnd As Integer)

        Dim lstLocations As List(Of udtProteinLocationInfo) = Nothing

        If mProteinMapInfo.TryGetValue(proteinName, lstLocations) Then
            ' Protein mapping already exists; check residueStart
            For Each udtLoc In lstLocations
                If udtLoc.ResidueStart = residueStart Then
                    ' Update this entry
                    If udtLoc.ResidueEnd <> residueEnd Then
                        udtLoc.ResidueEnd = residueEnd
                    End If
                    Exit Sub
                End If
            Next

            Dim udtLocInfoAddnl = New udtProteinLocationInfo
            udtLocInfoAddnl.ResidueStart = residueStart
            udtLocInfoAddnl.ResidueEnd = residueEnd

            lstLocations.Add(udtLocInfoAddnl)
            Exit Sub
        End If

        Dim udtLocInfo = New udtProteinLocationInfo
        udtLocInfo.ResidueStart = residueStart
        udtLocInfo.ResidueEnd = residueEnd

        mProteinMapInfo.Add(proteinName, New List(Of udtProteinLocationInfo) From {udtLocInfo})

    End Sub

End Class

File: PHRPReader/clsPepToProteinMapInfo.cs
using System;
using System.Collections.Generic;

namespace PHRPReader
{
    /// <summary>
    /// Track location of a given peptide in one or more proteins
    /// </summary>
    public class clsPepToProteinMapInfo
    {
        /// <summary>
        /// Start and end residue locations in a protein
        /// </summary>
        public class udtProteinLocationInfo
        {
            /// <summary>
            /// Start residue (first residue is 1)
            /// </summary>
            public int ResidueStart;

            /// <summary>
            /// End residue
            /// </summary>
            public int ResidueEnd;
        }

        /// <summary>
        /// Dictionary of protein names and residue start/end positions for a given peptide
        /// </summary>
        private readonly Dictionary<string, List<udtProteinLocationInfo>> mProteinMapInfo;

        /// <summary>
        /// Number of proteins that contain a given peptide
        /// </summary>
        public int ProteinCount => ProteinMapInfo.Count;

        /// <summary>
        /// Dictionary of protein names and residue start/end positions for a given peptide
        /// </summary>
        public IReadOnlyDictionary<string, List<udtProteinLocationInfo>> ProteinMapInfo => mProteinMapInfo;

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="proteinName">Protein name</param>
        /// <param name="residueStart">Location that a peptide starts in the protein</param>
        /// <param name="residueEnd">Location that a peptide ends in the protein</param>
        public clsPepToProteinMapInfo(string proteinName, int residueStart, int residueEnd)
        {
            mProteinMapInfo = new Dictionary<string, List<udtProteinLocationInfo>>(StringComparer.CurrentCultureIgnoreCase);

            AddProtein(proteinName, residueStart, residueEnd);
        }

        public void AddProtein(string proteinName, int residueStart, int residueEnd)
        {
            List<udtProteinLocationInfo> lstLocations = null;

            if (mProteinMapInfo.TryGetValue(proteinName, out lstLocations))
            {
                // Protein mapping already exists; check residueStart
                foreach (var udtLoc in lstLocations)
                {
                    if (udtLoc.ResidueStart == residueStart)
                    {
                        // Update this entry
                        if (udtLoc.ResidueEnd != residueEnd)
                        {
                            udtLoc.ResidueEnd = residueEnd;
                        }
                        return;
                    }
                }

                var udtLocInfoAddnl = new udtProteinLocationInfo();
                udtLocInfoAddnl.ResidueStart = residueStart;
                udtLocInfoAddnl.ResidueEnd = residueEnd;

                lstLocations.Add(udtLocInfoAddnl);
                return;
            }

            var udtLocInfo = new udtProteinLocationInfo();
            udtLocInfo.ResidueStart = residueStart;
            udtLocInfo.ResidueEnd = residueEnd;

            mProteinMapInfo.Add(proteinName, new List<udtProteinLocationInfo> { udtLocInfo });
        }
    }
}
